### PR TITLE
Telegram Bot API 3.1 Update

### PR DESCRIPTION
**Pull request by Olfi01**
I put some time in it but it wasn't that hard :smile: 
please, kindly accept my pull request :tongue:

### CODE REVIEW

**Comment by Olfi01** [comment]
@Olgabrezel Look and approve

**Comment by Olfi01** [comment]
@MrRoundRobin I'm terribly sorry for annoying you, but there are some people i know who are desperately waiting for this update 😓 sorry for the hassle

**Comment by FernandoTBarros** [comment]
YEES MEE!
Please aprove this!

**Comment by dllz** [comment]
I too am waiting for this update

**Comment by dllz** [comment]
Maybe if we say his name 3 times or something he will come @Olfi01?

**Comment by pouladpld** [comment]
I have created an organization and forked this project.
<https://github.com/TelegramBots/Telegram.Bot.Core>

Feel free to join there as collaborator and make your PR there. @MrRoundRobin can merge the code there into this repo. However, I am building `Telegram.Bot.Core` NuGet package out of that repository.

My goal is to get updates to the project faster. I also have setup integration tests and currently am working on testing Payments API.

**Comment by Olfi01** [comment]
@dllz We can try but i don't think it will work, @MrRoundRobin seems either too busy or @MrRoundRobin is just lazy or @MrRoundRobin s internet connection died. idk.

**Comment by Olfi01** [comment]
wait. he merged two pulls 3 days ago and just ignored this. now im feeling a bit angry.

**Comment by dllz** [comment]
Maybe he feels that 3.0 isnt tested yet on his side and doesnt wanna move to 3.1 yet? 
or he has another plan. If @MrRoundRobin could please let us know what is up that would be great. 
Thanks

**Comment by pouladpld** [comment]
Thanks to @Olfi01 👍  , I have tested and published [Telegram.Bot.Core](https://www.nuget.org/packages/Telegram.Bot.Core) package with support for API 3.1 methods.

**Comment by pouladpld** [comment]
I just changed a few little things such as converting`until_date` from `long` to `DateTime`(UTC).

File: src/Telegram.Bot/ITelegramBotClient.cs [hunk 1]
```diff
@@ -564,6 +565,45 @@ public interface ITelegramBotClient
...
+        /// <param name="canPinMessages">Pass True, if the administrator can pin messages, supergroups only</param>
+        /// <param name="canPromoteMembers">Pass True, if the administrator can add new administrators with a subset of his own privileges or demote administrators that he has promoted, directly or indirectly (promoted by administrators that were appointed by him)</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>Returns True on success.</returns>
+        /// <remarks>Pass False for all boolean parameters to demote a user.</remarks>
+        Task<bool> PromoteChatMemberAsync(ChatId chatId, int userId, bool? canChangeInfo = null,
+            bool? canPostMessages = null, bool? canEditMessages = null, bool? canDeleteMessages = null,
+            bool? canInviteUsers = null, bool? canRestrictMembers = null, bool? canPinMessages = null,
+            bool? canPromoteMembers = null, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/ITelegramBotClient.cs]: I would set the default for the boolean values to the same default as telegram.
I think the nullable type is overhead.

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 2]
```diff
@@ -1635,10 +1859,11 @@ public Task<GameHighScore[]> GetGameHighScoresAsync(int userId, string inlineMes
                 throw new ApiRequestException("Request timed out", 408, e);
             }
             catch (HttpRequestException e)
-                when (e.Message.Contains("400") || e.Message.Contains("403") || e.Message.Contains("409")) {}
+                when (e.Message.Contains("400") || e.Message.Contains("403") || e.Message.Contains("409"))
+            { }
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/TelegramBotClient.cs]: can you remove this unnecessary changes?

File: src/Telegram.Bot/Types/ChatMember.cs [hunk 3]
```diff
@@ -20,5 +20,89 @@ public class ChatMember
         /// </summary>
         [JsonProperty("status", Required = Required.Always)]
         public ChatMemberStatus Status { get; set; }
+
+        /// <summary>
+        /// Optional. Restictred and kicked only. Date when restrictions will be lifted for this user, unix time
+        /// </summary>
+        [JsonProperty(PropertyName = "until_date", Required = Required.Default)]
+        public int UntilDate { get; set; }
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/Types/ChatMember.cs]: this should be `DateTime`

File: src/Telegram.Bot/ITelegramBotClient.cs [hunk 4]
```diff
@@ -475,10 +475,11 @@ public interface ITelegramBotClient
         /// <param name="chatId"><see cref="ChatId"/> for the target group</param>
         /// <param name="userId">Unique identifier of the target user</param>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <param name="untilDate">Date when the user will be unbanned, unix time. If user is banned for more than 366 days or less than 30 seconds from the current time they are considered to be banned forever</param>
         /// <returns><c>true</c> on success.</returns>
         /// <see href="https://core.telegram.org/bots/api#kickchatmember"/>
         Task<bool> KickChatMemberAsync(ChatId chatId, int userId,
-            CancellationToken cancellationToken = default(CancellationToken));
+            CancellationToken cancellationToken = default(CancellationToken), int untilDate = 0);
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/ITelegramBotClient.cs]: - `untilDate` should be `DateTime` as indiated [here](https://github.com/MrRoundRobin/telegram.bot/pull/366#issuecomment-314290887)
- I always reordered the parameters so that the `cancellationToken` is the last. this caused a major change but I think this is ok for such a big change.

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 1 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: point is, i want people to be able to not change it at all, which would be just not passing the params, which would mean we'd need a "don't pass this" value, which would be null.
(holy s*** so many witches)

File: src/Telegram.Bot/TelegramBotClient.cs (hunk 2 above)
Comment (Olfi01) [inline src/Telegram.Bot/TelegramBotClient.cs]: Yea sorry, autoformat of visual studio

File: src/Telegram.Bot/Types/ChatMember.cs (hunk 3 above)
Comment (Olfi01) [inline src/Telegram.Bot/Types/ChatMember.cs]: will it parse the unix time automatically? i hope so

Comment (Olfi01) [inline src/Telegram.Bot/Types/ChatMember.cs]: apparently it won't. Gonna fix it in another way.

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 1 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: i just tested it, the default is indeed null, so we need nullables (i guess)

**Comment by Olfi01** [comment]
@MrRoundRobin dunno if the mention in the commit 3573c06 name reached you, so here I go again xD

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 4 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: darn i forgot first one, gonna do it

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 1 above)
Comment (MrRoundRobin) [inline src/Telegram.Bot/ITelegramBotClient.cs]: ah sorry is see, you'r right 👍

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 4 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: 58144e5

**Comment by Olfi01** [comment]
now im finished for real @MrRoundRobin xD

File: src/Telegram.Bot/Types/ChatMember.cs [hunk 5]
```diff
@@ -20,5 +21,94 @@ public class ChatMember
...
+        /// <summary>
+        /// Optional. Restictred and kicked only. Date when restrictions will be lifted for this user, unix time
+        /// </summary>
+        [JsonProperty(PropertyName = "until_date", Required = Required.Default)]
+        private long UntilDateUnix = 0;
+
+        /// <summary>
+        /// Optional. Restictred and kicked only. Date when restrictions will be lifted for this user, UTC time
+        /// </summary>
+        public DateTime UntilDate { get { return DateTime.FromFileTimeUtc(UntilDateUnix); } set { UntilDateUnix = value.ToFileTimeUtc(); } }
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/Types/ChatMember.cs]: this can be done without a private property: https://github.com/MrRoundRobin/telegram.bot/blob/a55f7dd66e0fa2be0edb5adeedbaa93ba60d31e9/src/Telegram.Bot/Types/Message.cs#L33

`[JsonConverter(typeof (UnixDateTimeConverter))]`

File: src/Telegram.Bot/TelegramBotClient.cs [hunk 6]
```diff
@@ -1007,6 +1010,92 @@ public Task<Chat> GetChatAsync(ChatId chatId, CancellationToken cancellationToke
...
+        /// <param name="canSendMessages">Pass True, if the user can send text messages, contacts, locations and venues</param>
+        /// <param name="canSendMediaMessages">Pass True, if the user can send audios, documents, photos, videos, video notes and voice notes, implies can_send_messages</param>
+        /// <param name="canSendOtherMessages">Pass True, if the user can send animations, games, stickers and use inline bots, implies can_send_media_messages</param>
+        /// <param name="canAddWebPagePreviews">Pass True, if the user may add web page previews to their messages, implies can_send_media_messages</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>On success, <c>true</c> is returned</returns>
+        /// <remarks>Pass True for all boolean parameters to lift restrictions from a user.</remarks>
+        public Task<bool> RestrictChatMemberAsync(ChatId chatId, int userId, int untilDate = 0,
```
Comment (MrRoundRobin) [inline src/Telegram.Bot/TelegramBotClient.cs]: this `untilDate` should also be an `DateTime`

File: src/Telegram.Bot/Types/ChatMember.cs (hunk 5 above)
Comment (Olfi01) [inline src/Telegram.Bot/Types/ChatMember.cs]: oh okay nice xD sorry

File: src/Telegram.Bot/TelegramBotClient.cs (hunk 6 above)
Comment (Olfi01) [inline src/Telegram.Bot/TelegramBotClient.cs]: oh god i hope its alright now @MrRoundRobin

**Comment by Olfi01** [comment]
oh god i hope its alright now @MrRoundRobin

**Comment by Olfi01** [comment]
Wait what do you Meran by the dateTimeConverter there? In the support Method?

File: src/Telegram.Bot/Types/ChatMember.cs (hunk 5 above)
Comment (Olfi01) [inline src/Telegram.Bot/Types/ChatMember.cs]: and how to use that for the DateTime s being passed to the Converter? @MrRoundRobin

**Comment by MrRoundRobin** [comment]
we need to tell Json.Net to convert datetime values to unix timestamps.

there are several ways, I think it's best to do this globaly. like this:
```csharp
JsonConvert.DefaultSettings = () => new JsonSerializerSettings
{
    Converters = new List<JsonConverter> { new SomeConverter() }
};
```
maybe this also reduces other code like the `FileToSend` handling. there was a reason why i din't done this before but can't remember 🤔  

I think it's a good idea to change something with my custom converters, I'll do this in the next two days.

if you add the missing `<see href="">` tags, i'll merge this pull request

**Comment by Olfi01** [comment]
@MrRoundRobin you meant those, right? 602a57a

**Comment by MrRoundRobin** [comment]
@Olfi01 [those](https://github.com/MrRoundRobin/telegram.bot/blob/master/src/Telegram.Bot/ITelegramBotClient.cs#L137)

(Need to add a contributing file 😃 )

**Comment by Olfi01** [comment]
oh. alright.

**Comment by Olfi01** [comment]
@MrRoundRobin contribution file? ugh i didnt read that screw me

**Comment by Olfi01** [comment]
oh i get it its something you gotta do phew xD

**Comment by MrRoundRobin** [comment]
now it looks nice, I'll do my best do merge it in the next 2 days

**Comment by Olfi01** [comment]
err you actually already have a UnixDateTimeConverter, you know? @MrRoundRobin 
couldn't we just add that one to JsonConvert.DefaultSettings.Converters ?

**Comment by MrRoundRobin** [comment]
yes, but i dont't know how this would impact projects referencing the library. i don't want to overrride their converters

**Comment by Olfi01** [comment]
nah i dont think it would...
should i revert anyway?
it works like this, i tested it @MrRoundRobin

**Comment by MrRoundRobin** [comment]
you could try it or I do it when i get time

**Comment by Olfi01** [comment]
it actually overrides them -.- ill see how you did it with `FileToSend` and do it the same

**Comment by Olfi01** [comment]
ah, i see... that won't be possible with DateTime, maybe add a custom class extending DateTime

**Comment by MrRoundRobin** [comment]
file to send is special, it will never be sent to telegram. we can use `JsonSerializerSettings`

**Comment by Olfi01** [comment]
`var payload = JsonConvert.SerializeObject(parameters, new UnixDateTimeConverter());`
shouldnt that work

**Comment by MrRoundRobin** [comment]
yes, but I think it woul make sense to add more converters, don't know if this is possible in this way

**Comment by MrRoundRobin** [comment]
@Olfi01 okay, didn't looked at the doku.

I think there are mor SerializeObject calls

**Comment by Olfi01** [comment]
None that could receive a DateTime. In the whole solution. @MrRoundRobin

File: src/Telegram.Bot/ITelegramBotClient.cs [hunk 7]
```diff
@@ -474,10 +474,11 @@ public interface ITelegramBotClient
         /// </summary>
         /// <param name="chatId"><see cref="ChatId"/> for the target group</param>
         /// <param name="userId">Unique identifier of the target user</param>
+        /// <param name="untilDate"><see cref="DateTime"/> when the user will be unbanned. If user is banned for more than 366 days or less than 30 seconds from the current time they are considered to be banned forever</param>
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
         /// <returns><c>true</c> on success.</returns>
         /// <see href="https://core.telegram.org/bots/api#kickchatmember"/>
-        Task<bool> KickChatMemberAsync(ChatId chatId, int userId,
+        Task<bool> KickChatMemberAsync(ChatId chatId, int userId, DateTime untilDate = default(DateTime),
```
Comment (pouladpld) [inline src/Telegram.Bot/ITelegramBotClient.cs]: I'm not really sure if this default value for `untilDate` is a good idea. I should be able to pass the following values and kick the user for ever:
```
DateTime.MinValue
DateTime.MaxValue
DateTime.UtcNow.AddSeconds(29)
DateTime.UtcNow.AddDays(366)
```

Also, is this properly parsed into UTC date time? Epoch time is based on UTC.

Here is how I resolved it: <https://github.com/TelegramBots/Telegram.Bot.Core/blob/master/src/Telegram.Bot/TelegramBotClient.cs#L848-L861>

And this is the test: <https://github.com/TelegramBots/Telegram.Bot.Core/blob/master/test/Telegram.Bot.Tests.Integ/AdminBots/ChatMemberAdministrationTests.cs#L139-L150>

**Comment by Olfi01** [comment]
@pouladpld you can pass Rhode and it will do exactly what youd want. No need for an fix.

**Comment by Olfi01** [comment]
Whoa autocorrect sigh
*any of those
*any

**Comment by Olfi01** [comment]
And yes, we've got an unixTimeConverter @pouladpld

**Comment by Olfi01** [comment]
So will you merge this @MrRoundRobin ?

File: src/Telegram.Bot/ITelegramBotClient.cs [hunk 8]
```diff
@@ -856,5 +898,82 @@ public interface ITelegramBotClient
...
+
+        /// <summary>
+        /// Use this method to change the description of a supergroup or a channel. The bot must be an administrator in the chat for this to work and must have the appropriate admin rights.
+        /// </summary>
+        /// <param name="chatId">Unique identifier for the target chat or username of the target channel</param>
+        /// <param name="description">New chat description, 0-255 characters</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>Returns true on success.</returns>
+        /// <see href="https://core.telegram.org/bots/api#setchatdescription"/>
+        Task<bool> SetChatDescriptionAsync(ChatId chatId, string description,
```
Comment (pouladpld) [inline src/Telegram.Bot/ITelegramBotClient.cs]: Description could be default to `null` and that means removing chat description.

<https://core.telegram.org/bots/api#setchatdescription>

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 7 above)
Comment (pouladpld) [inline src/Telegram.Bot/ITelegramBotClient.cs]: @Olfi01 If I always need to pass it values such as `DateTime.UtcNow`, it makes more sense to me to change parameter name to `utcUntilDate`. Instructs the clients to pass it as UTC time.

Comment (MrRoundRobin) [inline src/Telegram.Bot/ITelegramBotClient.cs]: @pouladpld I would reather change the converter to do the Utc part.

that's something I wanted to do a long time ago

**Comment by MrRoundRobin** [comment]
can you retarget this pr to the develop branch? i added the datetime requiremnts there already

**Comment by Olfi01** [comment]
I Gotta create a new PR for that right

**Comment by Olfi01** [comment]
I will close this since there's another PR to develop branch.

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 8 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: you're right, I'll do that too, idek why i made that different from official docs

Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: i did. It just defaults to an empty string instead and is then tested to be NullOrEmpty. Also, I forgot to add the default in ITelegramBotClient

File: src/Telegram.Bot/ITelegramBotClient.cs (hunk 7 above)
Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: @pouladpld you dont need to, you can just leave it out too

Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: and as @MrRoundRobin said, the converter should do the UTC part

Comment (Olfi01) [inline src/Telegram.Bot/ITelegramBotClient.cs]: i did that too. wasn't hard.
